File: solutions/78/main.py
class Solution:
    def __init__(self): 
        self.path = []
        self.result = []
    def backtracking(self, nums, startId): 
        # collect node 
        self.result.append(self.path[:])
        # base case 
        if startId == len(nums):
            return 
        for i in range(startId, len(nums)):
            self.path.append(nums[i])
            self.backtracking(nums, i+1)
            self.path.pop()
    def subsets(self, nums):
        self.backtracking(nums, 0)
        return self.result 

if __name__ == "__main__":
    s = Solution()
    nums = [1,2,3]
    print(s.subsets(nums))

File: solutions/78/main.cpp
/* ************************************************************************
> File Name:     main.cpp
> Author:        Sean
> Project:    LeetCode
> Created Time:  Wed Nov  3 17:01:43 2021
> Description:   
 ************************************************************************/
#include <vector> 
#include <gtest/gtest.h> 

using namespace std; 

class Solution {
private:
    vector<vector<int>> result;
    vector<int> path;
public:
    void backtracking(vector<int>& nums, int startId) {
        // record the node
        result.push_back(path);
        // base case
        if (startId == nums.size()) {
            return;
        }
        for (int i = startId; i < nums.size(); ++i) {
            path.push_back(nums[i]);
            backtracking(nums, i+1);
            path.pop_back();
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        backtracking(nums, 0);
        return result;
    }
};

TEST(Test78, SimpleTest) {
    vector<int> nums{
        1,2,3
    };
    Solution s; 
    vector<vector<int>> result = s.subsets(nums);
    for (auto& num : result) {
        cout << "subset ";
        for (auto& n : num) 
            cout << n << " ";
        cout << endl; 
    }
}

int main() {
    testing::InitGoogleTest();
    return RUN_ALL_TESTS();
}
